examples/amp-web-push/publisher-files/helper-frame.js

```diff
@@ -0,0 +1,277 @@
+     */
+    this.allowedWorkerMessageTopics = {};
+  }
+
+  /*
+    Ensures replies to the AMP page messenger have a consistent payload format.
+   */
+  replyToFrameWithPayload(
+    replyToFrameFunction,
+    wasSuccessful,
+    errorPayload,
+    successPayload) {
+    replyToFrameFunction({
+      success: wasSuccessful,
+      error: wasSuccessful ? undefined : errorPayload,
+      result: wasSuccessful ? successPayload : undefined
+    });
+  }
+
+  onAmpPageMessageReceived_NotificationPermissionState(_, replyToFrame) {
+    this.replyToFrameWithPayload(
+      replyToFrame,
+      true,
+      null,
+      Notification.permission
+    );
+  }
+
+  onAmpPageMessageReceived_ServiceWorkerState(_, replyToFrame) {
+    const serviceWorkerState = {
+      /*
+        Describes whether navigator.serviceWorker.controller is non-null.
+
+        A page hard-refreshed bypassing the cache will have the
+        navigator.serviceWorker.controller property be null even if the service
+        worker is successfully registered and activated. In these situations,
+        communication with the service worker isn't possible since the
+        controller is null.
+       */
+      isControllingFrame: !!navigator.serviceWorker.controller,
+      /*
+        The URL to the service worker script.
+       */
+      url: navigator.serviceWorker.controller ?
+        navigator.serviceWorker.controller.scriptURL :
+        null,
+      /*
+        The state of the service worker, one of "installing, waiting,
+        activating, activated".
+       */
+      state: navigator.serviceWorker.controller ?
+        navigator.serviceWorker.controller.state :
+        null
+    };
+
+    this.replyToFrameWithPayload(replyToFrame, true, null, serviceWorkerState);
+  }
+
+  onAmpPageMessageReceived_ServiceWorkerRegistration(message, replyToFrame) {
+    if (!message || !message.workerUrl || !message.registrationOptions) {
+      throw new Error('Expected arguments workerUrl and registrationOptions ' +
+      'in message, got:', message);
+    }
+
+    navigator.serviceWorker.register(
+      message.workerUrl,
+      message.registrationOptions
+    ).then(registration => {
+        this.replyToFrameWithPayload(replyToFrame, true, null, null);
+      })
+      .catch(error => {
+        this.replyToFrameWithPayload(replyToFrame, true, null, error ?
+          (error.message || error.toString()) :
+          null
+        );
+      });
+  }
+
+  onAmpPageMessageReceived_ServiceWorkerQuery(message, replyToFrame) {
+    if (!message || !message.topic) {
+      throw new Error('Expected argument topic in message, got:', message);
+    }
+    new Promise((resolve) => {
+      // Allow this message through, just for the next time it's received
+      this.allowedWorkerMessageTopics[message.topic] = resolve;
+
+      // The AMP message is forwarded to the service worker
+      return this.waitUntilWorkerControlsPage().then(() => {
```

Comment:
I removed the unnecessary return, but this Promise resolves only when `resolve()` is called (which is stored to be called later). So the return statements inside this `new Promise() { .. }` don't affect the resolution of the Promise.